### PR TITLE
Generic things: Add extended smart meter

### PR DESCRIPTION
**Pull request by Boernsman**
nymea-plugins pull request checklist:

- [x] Make sure the pull request's title is of format "Plugin name: Add support for xyz" or "New plugin: Plugin name"

- [x] Did you test the changes on hardware, if not (e.g. absence of required hardware), please mention a person to confirm it has been tested.

- [x] Did you update the plugin's README.md accordingly?

- [x] Did you update translations (`cd builddir && make lupdate`)?

- [x] If you added a new plugin, should it be added to nymea-plugins-all or nymea-plugins-maker?

### CODE REVIEW

File: genericthings/integrationplugingenericthings.json [hunk 1]
```diff
@@ -693,6 +693,55 @@
...
+                    "id": "c3123967-f741-4fe1-a0d4-9a3e405d7e52",
+                    "name": "extendedSmartMeterConsumer",
+                    "displayName": "Extended smart meter consumer",
+                    "createMethods": ["user"],
+                    "interfaces": ["extendedsmartmeterconsumer"],
+                    "settingsTypes": [
+                        {
+                            "id": "c361732b-68eb-447e-a434-e84031231871",
+                            "name": "impulsePerKwh",
+                            "displayName": "Impulse/kWh",
```
Comment (mzanetti) [inline genericthings/integrationplugingenericthings.json:707]: Impulses, plural, no?

File: genericthings/integrationplugingenericthings.json [hunk 2]
```diff
@@ -693,6 +693,55 @@
                             "ioType": "analogOutput"
                         }
                     ]
+                },
+                {
+                    "id": "c3123967-f741-4fe1-a0d4-9a3e405d7e52",
+                    "name": "extendedSmartMeterConsumer",
+                    "displayName": "Extended smart meter consumer",
```
Comment (mzanetti) [inline genericthings/integrationplugingenericthings.json:700]: That's a very technical description. Wouldn't be something like "Impulse based energy meter" better for the user?

File: genericthings/integrationplugingenericthings.json [hunk 3]
```diff
@@ -693,6 +693,55 @@
...
+                            "displayName": "Energy",
+                            "displayNameEvent": "Energy changed",
+                            "type": "double",
+                            "defaultValue": 0.00,
+                            "unit": "KiloWattHour"
+                        },
+                        {
+                            "id": "9cd7e5ca-f8f8-48d5-9785-911ae75158c3",
+                            "name": "s0Input",
+                            "displayName": "S0 input",
```
Comment (mzanetti) [inline genericthings/integrationplugingenericthings.json:736]: What's S0?

File: genericthings/integrationplugingenericthings.cpp [hunk 4]
```diff
@@ -490,6 +512,13 @@ void IntegrationPluginGenericThings::thingRemoved(Thing *thing)
         m_extendedBlindTargetPercentage.remove(thing);
         m_venetianBlindAngleTimer.take(thing)->deleteLater();
         m_venetianBlindTargetAngle.remove(thing);
+    } else if (thing->thingClassId() == extendedSmartMeterConsumerThingClassId) {
+        m_pulsesPerTimeframe.remove(thing);
+    }
+
+    if (myThings().isEmpty()) {
```
Comment (mzanetti) [inline genericthings/integrationplugingenericthings.cpp:519]: should filter for the appropriate thing class

File: genericthings/integrationplugingenericthings.cpp [hunk 5]
```diff
@@ -475,6 +482,21 @@ void IntegrationPluginGenericThings::executeAction(ThingActionInfo *info)
...
         }
+    } else if (thing->thingClassId() == extendedSmartMeterConsumerThingClassId) {
+        if (action.actionTypeId() == extendedSmartMeterConsumerS0InputActionTypeId) {
+            bool value = info->action().param(extendedSmartMeterConsumerS0InputActionS0InputParamTypeId).value().toBool();
+            thing->setStateValue(extendedSmartMeterConsumerS0InputStateTypeId, value);
+            int impulsePerKwh = info->thing()->setting(extendedSmartMeterConsumerSettingsImpulsePerKwhParamTypeId).toInt();
+            if (value) {
+                double currentEnergy = thing->stateValue(extendedSmartMeterConsumerTotalEnergyConsumedStateTypeId).toDouble();
+                thing->setStateValue(extendedSmartMeterConsumerTotalEnergyConsumedStateTypeId ,currentEnergy + (1.00/impulsePerKwh));
+                m_pulsesPerTimeframe.insert(thing, m_pulsesPerTimeframe.value(thing)+1);
```
Comment (mzanetti) [inline genericthings/integrationplugingenericthings.cpp:493]: you can/should probably use this instead:
`m_pulsesPerTimeframe[thing]++`
Note that the operator[] will pass a reference to the value which can be modified (unlike .value() which will return a copy of the value)

Comment (mzanetti) [reply genericthings/integrationplugingenericthings.cpp:493]: (also, operator[] will insert if not existing, so no need to worry about acessing something not existing)

File: genericthings/integrationplugingenericthings.json (hunk 1 above)
Comment (Boernsman) [reply genericthings/integrationplugingenericthings.json:707]: usually it is revered as impulse/kwh or short imp/kwh

**Pull request merged**